In [1]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style('whitegrid')

In [2]:
# Allow display of all columns
pd.set_option('display.max_columns', None)

# loading the data
nmes_df = pd.read_csv("nmes.txt", sep="\t", header=0)
nmes_df.head(10)

PIDX       LASTAGE   MALE   RACE3   eversmk   current   former   AGESMOKE   \
0   20843014        80      0       3  0         .         0        .           
1   20833013        78      0       3  .         .         .        .           
2   20832010        80      1       3  1         1         0        21          
3   20829011        71      1       3  1         0         1        16          
4   20829024        69      0       3  .         .         .        .           
5   20827015        70      0       3  .         .         .        .           
6   20825018        60      0       3  1         0         1        30          
7   20821016        82      0       3  1         0         1        16          
8   20820013        58      1       3  .         .         .        .           
9   20820026        46      0       3  .         .         .        .           

   CIGSSMOK   CIGSADAY   AGESTOP   packyears   yearsince   INCALPER       \
0  .          .          .         0           0             5805.555102   
1  .          .          .         .           .             7428.738855   
2  10         10         .         30          0             8220.839822   
3  .          .          46        .           25            4857.094656   
4  .          .          .         .           .             4808.212619   
5  .          .          .         .           .             5663.136054   
6  20         .          38        9           22           12289.275970   
7  2          .          35        2           47            8790.163608   
8  .          .          .         .           .            14242.308460   
9  .          .          .         .           .            15434.187510   

   TOTALEXP    lc5   chd5   beltuse   educate   marital   SREGION  POVSTALB  
0      297.75     0      1  3                4  2                1        1  
1      451.57     0      0  3                3  2                1        4  
2     1368.76     0      1  1                3  3                1        3  
3     1205.32     0      0  3                3  1                1        5  
4      662.02     0      0  .                3  1                1        5  
5      286.97     0      0  .                1  2                1        4  
6      795.00     0      0  3                2  5                1        5  
7     1627.09     0      0  3                3  5                1        4  
8        0.00     0      0  .                1  1                1        3  
9      110.00     0      0  .                1  1                1        3

In [3]:
nmes_df.sample(10)

PIDX       LASTAGE   MALE   RACE3   eversmk   current   former   \
12665   37201022        65      0       3  0         .         0         
5211    27184018        77      0       3  0         .         0         
7636    24230028        60      0       3  0         .         0         
8559    23841019        68      1       3  0         .         0         
793     29770016        42      0       2  .         .         .         
4027    36892027        49      0       3  0         .         0         
5347    27012011        40      1       2  0         .         0         
6274    25796020        74      0       3  0         .         0         
9779    35053018        50      1       1  1         0         1         
3051    26542017        41      0       3  1         0         1         

       AGESMOKE   CIGSSMOK   CIGSADAY   AGESTOP   packyears   yearsince   \
12665  .          .          .          .         0           0            
5211   .          .          .          .         0           0            
7636   .          .          .          .         0           0            
8559   .          .          .          .         0           0            
793    .          .          .          .         .           .            
4027   .          .          .          .         0           0            
5347   .          .          .          .         0           0            
6274   .          .          .          .         0           0            
9779   .          .          .          28        .           22           
3051   16         10         .          34        9.5         7            

       INCALPER       TOTALEXP    lc5   chd5   beltuse   educate   marital   \
12665    4577.809169        0.00     0      0  1                3  1          
5211     4396.497544      295.80     0      1  3                4  2          
7636    12881.569550     1194.32     0      0  3                2  1          
8559     3926.429229      402.15     0      0  3                1  1          
793      3295.261494    17505.86     0      0  .                1  .          
4027     9477.215058        0.00     0      0  2                3  4          
5347     4378.211378      123.75     0      0  3                1  1          
6274     3582.675580      385.02     0      0  3                3  1          
9779     5128.972566       32.00     0      0  1                4  4          
3051     6546.456217     1916.01     0      0  3                2  4          

       SREGION  POVSTALB  
12665         3        1  
5211          2        1  
7636          4        5  
8559          4        5  
793           3        3  
4027          3        5  
5347          3        4  
6274          2        5  
9779          2        4  
3051          4        4

In [4]:
nmes_df.tail(10)

PIDX       LASTAGE   MALE   RACE3   eversmk   current   former   \
13638   31379018        81      1       3  1         0         1         
13639   31379021        77      0       3  0         .         0         
13640   31378015        76      1       3  1         0         1         
13641   31377012        84      0       3  0         .         0         
13642   31374013        71      1       3  1         0         1         
13643   31372016        55      0       3  1         1         0         
13644   31370010        54      1       3  1         0         1         
13645   31370023        53      0       3  1         0         1         
13646   31368014        65      1       3  0         .         0         
13647   31368027        56      0       3  0         .         0         

       AGESMOKE   CIGSSMOK   CIGSADAY   AGESTOP   packyears   yearsince   \
13638  20         10         .          45        13          36           
13639  .          .          .          .         0           0            
13640  17         10         .          70        27          6            
13641  .          .          .          .         0           0            
13642  12         20         .          40        29          31           
13643  18         40         40         .         76          0            
13644  18         20         .          27        10          27           
13645  18         20         .          52        35          1            
13646  .          .          .          .         0           0            
13647  .          .          .          .         0           0            

       INCALPER       TOTALEXP    lc5   chd5   beltuse   educate   marital   \
13638    5055.967393      751.86     0      0  1                4  1          
13639    5634.858425        0.00     0      0  1                2  1          
13640    4579.520095     1154.00     0      0  2                4  2          
13641    6002.890503     6873.85     0      0  1                4  2          
13642    5553.125809     1469.93     0      1  3                3  2          
13643    9372.662007      115.00     0      0  2                2  3          
13644    9909.577179     1350.95     0      0  1                3  1          
13645    9951.748849     1303.90     0      0  2                3  1          
13646   11508.257920      387.90     0      0  3                2  1          
13647   12837.717840      309.00     0      0  3                3  1          

       SREGION  POVSTALB  
13638         4        4  
13639         4        4  
13640         4        4  
13641         4        3  
13642         4        4  
13643         4        4  
13644         4        5  
13645         4        5  
13646         4        5  
13647         4        5

## Data Understanding

In [5]:
nmes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13648 entries, 0 to 13647
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PIDX           13648 non-null  int64  
 1   LASTAGE        13648 non-null  int64  
 2   MALE           13648 non-null  int64  
 3   RACE3          13648 non-null  int64  
 4   eversmk        13648 non-null  object 
 5   current        13648 non-null  object 
 6   former         13648 non-null  object 
 7   AGESMOKE       13648 non-null  object 
 8   CIGSSMOK       13648 non-null  object 
 9   CIGSADAY       13648 non-null  object 
 10  AGESTOP        13648 non-null  object 
 11  packyears      13648 non-null  object 
 12  yearsince      13648 non-null  object 
 13  INCALPER       13648 non-null  float64
 14  TOTALEXP       13648 non-null  float64
 15  lc5            13648 non-null  int64  
 16  chd5           13648 non-null  int64  
 17  beltuse        13648 non-null  object 
 18  educat

In [6]:
nmes_df.shape

(13648, 22)

In [7]:
nmes_df.describe()

PIDX           LASTAGE          MALE         RACE3   INCALPER       \
count  1.364800e+04  13648.000000  13648.000000  13648.000000   13648.000000   
mean   2.888199e+07     60.653502      0.431419      2.694387    6520.868624   
std    5.110010e+06     13.198304      0.495292      0.589737    3274.185570   
min    2.000202e+07     40.000000      0.000000      1.000000     780.924047   
25%    2.444152e+07     49.000000      0.000000      3.000000    4061.730219   
50%    2.882252e+07     61.000000      0.000000      3.000000    5451.788234   
75%    3.331527e+07     71.000000      1.000000      3.000000    9152.403095   
max    3.761402e+07     94.000000      1.000000      3.000000   26912.163450   

          TOTALEXP            lc5          chd5       educate       SREGION   
count   13648.000000  13648.000000  13648.000000  13648.000000  13648.000000  
mean     2956.329748      0.020516      0.097230      2.829499      2.534804  
std      8393.412429      0.141762      0.296282      0.938898      1.007682  
min         0.000000      0.000000      0.000000      1.000000      1.000000  
25%       161.000000      0.000000      0.000000      2.000000      2.000000  
50%       600.000000      0.000000      0.000000      3.000000      3.000000  
75%      1878.795000      0.000000      0.000000      3.000000      3.000000  
max    175095.970000      1.000000      1.000000      4.000000      4.000000

In [8]:
nmes_df.dtypes

PIDX               int64
LASTAGE            int64
MALE               int64
RACE3              int64
eversmk           object
current           object
former            object
AGESMOKE          object
CIGSSMOK          object
CIGSADAY          object
AGESTOP           object
packyears         object
yearsince         object
INCALPER         float64
TOTALEXP         float64
lc5                int64
chd5               int64
beltuse           object
educate            int64
marital           object
SREGION            int64
POVSTALB          object
dtype: object

## Data Cleaning

### 1. Checking for Completeness

In [9]:
# checking for missing values
nmes_df.isna().sum()

PIDX             0
LASTAGE          0
MALE             0
RACE3            0
eversmk          0
current          0
former           0
AGESMOKE         0
CIGSSMOK         0
CIGSADAY         0
AGESTOP          0
packyears        0
yearsince        0
INCALPER         0
TOTALEXP         0
lc5              0
chd5             0
beltuse          0
educate          0
marital          0
SREGION          0
POVSTALB         0
dtype: int64

The Dataset seems to have no missing values, however , we know that the missing values are represented by '.' in the dataset. So we will replace '.' with NaN and then check for missing values.

In [10]:
nmes_df.head(10)

PIDX       LASTAGE   MALE   RACE3   eversmk   current   former   AGESMOKE   \
0   20843014        80      0       3  0         .         0        .           
1   20833013        78      0       3  .         .         .        .           
2   20832010        80      1       3  1         1         0        21          
3   20829011        71      1       3  1         0         1        16          
4   20829024        69      0       3  .         .         .        .           
5   20827015        70      0       3  .         .         .        .           
6   20825018        60      0       3  1         0         1        30          
7   20821016        82      0       3  1         0         1        16          
8   20820013        58      1       3  .         .         .        .           
9   20820026        46      0       3  .         .         .        .           

   CIGSSMOK   CIGSADAY   AGESTOP   packyears   yearsince   INCALPER       \
0  .          .          .         0           0             5805.555102   
1  .          .          .         .           .             7428.738855   
2  10         10         .         30          0             8220.839822   
3  .          .          46        .           25            4857.094656   
4  .          .          .         .           .             4808.212619   
5  .          .          .         .           .             5663.136054   
6  20         .          38        9           22           12289.275970   
7  2          .          35        2           47            8790.163608   
8  .          .          .         .           .            14242.308460   
9  .          .          .         .           .            15434.187510   

   TOTALEXP    lc5   chd5   beltuse   educate   marital   SREGION  POVSTALB  
0      297.75     0      1  3                4  2                1        1  
1      451.57     0      0  3                3  2                1        4  
2     1368.76     0      1  1                3  3                1        3  
3     1205.32     0      0  3                3  1                1        5  
4      662.02     0      0  .                3  1                1        5  
5      286.97     0      0  .                1  2                1        4  
6      795.00     0      0  3                2  5                1        5  
7     1627.09     0      0  3                3  5                1        4  
8        0.00     0      0  .                1  1                1        3  
9      110.00     0      0  .                1  1                1        3

In [11]:
nmes_df.replace(r'\.', np.nan, regex=True, inplace=True)
nmes_df.to_csv('nmes3.csv', index=False)
nmes_df.head(10)

PIDX       LASTAGE   MALE   RACE3   eversmk   current   former   AGESMOKE   \
0   20843014        80      0       3  0              NaN  0              NaN   
1   20833013        78      0       3       NaN       NaN      NaN        NaN   
2   20832010        80      1       3  1         1         0        21          
3   20829011        71      1       3  1         0         1        16          
4   20829024        69      0       3       NaN       NaN      NaN        NaN   
5   20827015        70      0       3       NaN       NaN      NaN        NaN   
6   20825018        60      0       3  1         0         1        30          
7   20821016        82      0       3  1         0         1        16          
8   20820013        58      1       3       NaN       NaN      NaN        NaN   
9   20820026        46      0       3       NaN       NaN      NaN        NaN   

   CIGSSMOK   CIGSADAY   AGESTOP   packyears   yearsince   INCALPER       \
0        NaN        NaN       NaN  0           0             5805.555102   
1        NaN        NaN       NaN         NaN         NaN    7428.738855   
2  10         10              NaN  30          0             8220.839822   
3        NaN        NaN  46               NaN  25            4857.094656   
4        NaN        NaN       NaN         NaN         NaN    4808.212619   
5        NaN        NaN       NaN         NaN         NaN    5663.136054   
6  20               NaN  38        9           22           12289.275970   
7  2                NaN  35        2           47            8790.163608   
8        NaN        NaN       NaN         NaN         NaN   14242.308460   
9        NaN        NaN       NaN         NaN         NaN   15434.187510   

   TOTALEXP    lc5   chd5   beltuse   educate   marital   SREGION  POVSTALB  
0      297.75     0      1  3                4  2                1        1  
1      451.57     0      0  3                3  2                1        4  
2     1368.76     0      1  1                3  3                1        3  
3     1205.32     0      0  3                3  1                1        5  
4      662.02     0      0       NaN         3  1                1        5  
5      286.97     0      0       NaN         1  2                1        4  
6      795.00     0      0  3                2  5                1        5  
7     1627.09     0      0  3                3  5                1        4  
8        0.00     0      0       NaN         1  1                1        3  
9      110.00     0      0       NaN         1  1                1        3

In [12]:
nmes_df.sample(20)

PIDX       LASTAGE   MALE   RACE3   eversmk   current   former   \
11084   31698020        58      1       3       NaN       NaN      NaN   
2395    23316011        77      1       3  1         0         1         
6562    32305033        51      0       3       NaN       NaN      NaN   
8505    23908014        57      1       1       NaN       NaN      NaN   
10922   30204025        55      0       3  1         1         0         
7864    33922010        73      0       3  1              NaN      NaN   
10226   24604028        49      0       2  1         0         1         
807     29756013        72      1       3  1         0         1         
9225    35744014        53      1       1  0              NaN  0         
1985    24884011        81      0       3  0              NaN  0         
6988    30002017        66      1       3  0              NaN  0         
220     22020023        75      0       3  0              NaN  0         
7173    34252016        64      1       3  1         0         1         
2705    25492019        83      1       3  0              NaN  0         
13408   22492010        69      0       3  1         0         1         
2358    23346011        57      1       3       NaN       NaN      NaN   
5373    27983013        65      1       2       NaN       NaN      NaN   
11784   27445017        67      1       3  1         0         1         
11785   27445020        56      0       3  0              NaN  0         
8204    20875012        50      1       2  1         0         1         

       AGESMOKE   CIGSSMOK   CIGSADAY   AGESTOP   packyears   yearsince   \
11084        NaN        NaN        NaN       NaN         NaN         NaN   
2395   20         20               NaN  77        58          0            
6562         NaN        NaN        NaN       NaN         NaN         NaN   
8505         NaN        NaN        NaN       NaN         NaN         NaN   
10922  19         18         18              NaN         NaN  0            
7864         NaN        NaN        NaN       NaN         NaN         NaN   
10226        NaN  10               NaN  32               NaN  17           
807    16         20               NaN  62        47          10           
9225         NaN        NaN        NaN       NaN  0           0            
1985         NaN        NaN        NaN       NaN  0           0            
6988         NaN        NaN        NaN       NaN  0           0            
220          NaN        NaN        NaN       NaN  0           0            
7173   16         40               NaN  47        64          17           
2705         NaN        NaN        NaN       NaN  0           0            
13408  34         60               NaN  53        60          16           
2358         NaN        NaN        NaN       NaN         NaN         NaN   
5373         NaN        NaN        NaN       NaN         NaN         NaN   
11784  22         10               NaN  30               NaN  37           
11785        NaN        NaN        NaN       NaN  0           0            
8204   14         10               NaN       NaN         NaN         NaN   

       INCALPER       TOTALEXP    lc5   chd5   beltuse   educate   marital   \
11084    5370.676715        0.00     0      0       NaN         4  2          
2395     4961.697603      981.80     0      0  2                4  1          
6562     5615.258568     8279.07     0      0       NaN         4  2          
8505     3911.803159      685.00     0      0  2                4  1          
10922    5131.431065      654.07     0      0  2                3  1          
7864     8673.292435     1385.00     0      0  3                3  2          
10226    3724.254728      290.35     0      0  2                3  4          
807      5691.545223     1421.04     1      0  3                3  4          
9225     9528.435960      129.24     0      0  2                4  5          
1985     4681.353382    13451.25     1      0  3                1

In [13]:
# checking for missing values again after replacing . with NaN
missing_values = nmes_df.isna().sum()
percent_missing = nmes_df.isna().sum()/nmes_df.shape[0]*100

value = {
    'missing_values ':missing_values,
    'percent_missing %':percent_missing
}
frame = pd.DataFrame(value)
frame

missing_values   percent_missing %
PIDX                         0           0.000000
LASTAGE                      0           0.000000
MALE                         0           0.000000
RACE3                        0           0.000000
eversmk                   1964          14.390387
current                   7211          52.835580
former                    2152          15.767878
AGESMOKE                  7407          54.271688
CIGSSMOK                  7944          58.206331
CIGSADAY                 10834          79.381594
AGESTOP                  10508          76.992966
packyears                 5088          37.280188
yearsince                 2462          18.039273
INCALPER                     0           0.000000
TOTALEXP                     0           0.000000
lc5                          0           0.000000
chd5                         0           0.000000
beltuse                   1243           9.107562
educate                      0           0.000000
marital                    320           2.344666
SREGION                      0           0.000000
POVSTALB                    56           0.410317

Based on the above analysis :
- The following columns have missing values above 50% : 

    - `current`  : 7211 missing values with 52%
    - `AGESMOKE` : 7407 missing values with 54%
    - `CIGSSMOK` : 7944	missing values with 58%
    - `CIGSADAY` : 10834 missing values with 79%
    - `AGESTOP`  : 10508 missing values with 77%
    

In [14]:
# dropping columns with more than 50% missing values
nmes_df = nmes_df.dropna(thresh=0.5*len(nmes_df), axis=1)

In [15]:
nmes_df.shape

(13648, 17)

In [16]:
nmes_df.isna().sum()

PIDX                0
LASTAGE             0
MALE                0
RACE3               0
eversmk          1964
former           2152
packyears        5088
yearsince        2462
INCALPER            0
TOTALEXP            0
lc5                 0
chd5                0
beltuse          1243
educate             0
marital           320
SREGION             0
POVSTALB           56
dtype: int64

In [17]:
nmes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13648 entries, 0 to 13647
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PIDX           13648 non-null  int64  
 1   LASTAGE        13648 non-null  int64  
 2   MALE           13648 non-null  int64  
 3   RACE3          13648 non-null  int64  
 4   eversmk        11684 non-null  object 
 5   former         11496 non-null  object 
 6   packyears      8560 non-null   object 
 7   yearsince      11186 non-null  object 
 8   INCALPER       13648 non-null  float64
 9   TOTALEXP       13648 non-null  float64
 10  lc5            13648 non-null  int64  
 11  chd5           13648 non-null  int64  
 12  beltuse        12405 non-null  object 
 13  educate        13648 non-null  int64  
 14  marital        13328 non-null  object 
 15  SREGION        13648 non-null  int64  
 16  POVSTALB       13592 non-null  object 
dtypes: float64(2), int64(8), object(7)
memory usage: 1

In [19]:
nmes_df.columns

Index(['PIDX     ', 'LASTAGE ', 'MALE ', 'RACE3 ', 'eversmk ', 'former ',
       'packyears ', 'yearsince ', 'INCALPER     ', 'TOTALEXP  ', 'lc5 ',
       'chd5 ', 'beltuse ', 'educate ', 'marital ', 'SREGION ', 'POVSTALB'],
      dtype='object')

In [22]:
nmes_df.head(10)

PIDX       LASTAGE   MALE   RACE3   eversmk   former   packyears   \
0   20843014        80      0       3  0         0        0            
1   20833013        78      0       3       NaN      NaN         NaN   
2   20832010        80      1       3  1         0        30           
3   20829011        71      1       3  1         1               NaN   
4   20829024        69      0       3       NaN      NaN         NaN   
5   20827015        70      0       3       NaN      NaN         NaN   
6   20825018        60      0       3  1         1        9            
7   20821016        82      0       3  1         1        2            
8   20820013        58      1       3       NaN      NaN         NaN   
9   20820026        46      0       3       NaN      NaN         NaN   

   yearsince   INCALPER       TOTALEXP    lc5   chd5   beltuse   educate   \
0  0             5805.555102      297.75     0      1  3                4   
1         NaN    7428.738855      451.57     0      0  3                3   
2  0             8220.839822     1368.76     0      1  1                3   
3  25            4857.094656     1205.32     0      0  3                3   
4         NaN    4808.212619      662.02     0      0       NaN         3   
5         NaN    5663.136054      286.97     0      0       NaN         1   
6  22           12289.275970      795.00     0      0  3                2   
7  47            8790.163608     1627.09     0      0  3                3   
8         NaN   14242.308460        0.00     0      0       NaN         1   
9         NaN   15434.187510      110.00     0      0       NaN         1   

   marital   SREGION  POVSTALB  
0  2                1        1  
1  2                1        4  
2  3                1        3  
3  1                1        5  
4  1                1        5  
5  2                1        4  
6  5                1        5  
7  5                1        4  
8  1                1        3  
9  1                1        3

In [26]:
# Replacing the missing values with a describe word "missing" for the categorical variables
nmes_df['eversmk '].fillna('missing', inplace=True)
nmes_df['former '].fillna('missing', inplace=True)
nmes_df['beltuse '].fillna('missing', inplace=True)
nmes_df['marital '].fillna('missing', inplace=True)
nmes_df['POVSTALB'].fillna('missing', inplace=True)

In [28]:
# Replacing the missing values with the mean for the numerical variables
nmes_df['packyears '] = nmes_df['packyears '].astype(float)
nmes_df['packyears '].fillna(nmes_df['packyears '].mean(), inplace=True)

nmes_df['yearsince '] = nmes_df['yearsince '].astype(float)
nmes_df['yearsince '].fillna(nmes_df['yearsince '].mean(), inplace=True)

In [29]:
# confirming that there are no missing values
nmes_df.isna().sum()

PIDX             0
LASTAGE          0
MALE             0
RACE3            0
eversmk          0
former           0
packyears        0
yearsince        0
INCALPER         0
TOTALEXP         0
lc5              0
chd5             0
beltuse          0
educate          0
marital          0
SREGION          0
POVSTALB         0
dtype: int64

### 2. Checking for Uniformity

In [30]:
nmes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13648 entries, 0 to 13647
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PIDX           13648 non-null  int64  
 1   LASTAGE        13648 non-null  int64  
 2   MALE           13648 non-null  int64  
 3   RACE3          13648 non-null  int64  
 4   eversmk        13648 non-null  object 
 5   former         13648 non-null  object 
 6   packyears      13648 non-null  float64
 7   yearsince      13648 non-null  float64
 8   INCALPER       13648 non-null  float64
 9   TOTALEXP       13648 non-null  float64
 10  lc5            13648 non-null  int64  
 11  chd5           13648 non-null  int64  
 12  beltuse        13648 non-null  object 
 13  educate        13648 non-null  int64  
 14  marital        13648 non-null  object 
 15  SREGION        13648 non-null  int64  
 16  POVSTALB       13648 non-null  object 
dtypes: float64(4), int64(8), object(5)
memory usage: 1

In [35]:
# changing the MALE column name to gender 
nmes_df = nmes_df.rename(columns={'MALE ': 'gender'})
nmes_df = nmes_df.rename(columns={'LASTAGE ': 'age'})
nmes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13648 entries, 0 to 13647
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PIDX           13648 non-null  int64  
 1   age            13648 non-null  int64  
 2   gender         13648 non-null  int64  
 3   RACE3          13648 non-null  int64  
 4   eversmk        13648 non-null  object 
 5   former         13648 non-null  object 
 6   packyears      13648 non-null  float64
 7   yearsince      13648 non-null  float64
 8   INCALPER       13648 non-null  float64
 9   TOTALEXP       13648 non-null  float64
 10  lc5            13648 non-null  int64  
 11  chd5           13648 non-null  int64  
 12  beltuse        13648 non-null  object 
 13  educate        13648 non-null  int64  
 14  marital        13648 non-null  object 
 15  SREGION        13648 non-null  int64  
 16  POVSTALB       13648 non-null  object 
dtypes: float64(4), int64(8), object(5)
memory usage: 1

In [36]:
# converting the column from int to category
nmes_df['gender'] = nmes_df['gender'].astype('category')
nmes_df['RACE3 '] = nmes_df['RACE3 '].astype('category')

In [37]:
# converting all the object data types to category
def convert_to_category(df):
    for column in df.columns:
        if df[column].dtype == 'object':
            df[column] = df[column].astype('category')
    return df

nmes_df = convert_to_category(nmes_df)

In [38]:
nmes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13648 entries, 0 to 13647
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   PIDX           13648 non-null  int64   
 1   age            13648 non-null  int64   
 2   gender         13648 non-null  category
 3   RACE3          13648 non-null  category
 4   eversmk        13648 non-null  category
 5   former         13648 non-null  category
 6   packyears      13648 non-null  float64 
 7   yearsince      13648 non-null  float64 
 8   INCALPER       13648 non-null  float64 
 9   TOTALEXP       13648 non-null  float64 
 10  lc5            13648 non-null  int64   
 11  chd5           13648 non-null  int64   
 12  beltuse        13648 non-null  category
 13  educate        13648 non-null  int64   
 14  marital        13648 non-null  category
 15  SREGION        13648 non-null  int64   
 16  POVSTALB       13648 non-null  category
dtypes: category(7), float64(4), int

### 3. Checking for Consistency

In [ ]:
# checking for duplicates

In [ ]:
nmes_df.duplicated().sum()

In [ ]:
# check if there any duplicated data for train
duplicate_1 = nmes_df[nmes_df.duplicated()]
if duplicate_1.empty== True:
    print("no duplicated data")
else:
    print("Duplicate Rows except first occurrence based on all columns are :")
    print(duplicate_1)

### 4. Checking for Validity

In [ ]:
# checking for outliers
main_df.boxplot(figsize=(12,10))

In [ ]:
# dealing with outliers
def check_deal_outlier(df, col):
    q1 = df[col].quantile(0.25)
    q3 = df[col].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr)
    upper_bound = q3 + (1.5 * iqr)
    print('Lower bound: ', lower_bound)
    print('Upper bound: ', upper_bound)
    print('Number of outliers: ', df[(df[col] < lower_bound) | (df[col] > upper_bound)].shape[0])
    print('Percentage of outliers: ', df[(df[col] < lower_bound) | (df[col] > upper_bound)].shape[0] / df.shape[0])
    print('Number of values within bounds: ', df[(df[col] >= lower_bound) & (df[col] <= upper_bound)].shape[0])
    print('Percentage of values within bounds: ', df[(df[col] >= lower_bound) & (df[col] <= upper_bound)].shape[0] / df.shape[0])
    return df[(df[col] >= lower_bound) & (df[col] <= upper_bound)]

main_df = check_deal_outlier(main_df, 'cc_num')

## Exploratory Data Analysis

## Data Preparation
- encoding categorical variables
- scaling numerical variables

In [ ]:
## Data Preparation
